CONTRIBUTING.md

```diff
@@ -141,19 +141,22 @@ tracing) enabled.
 
 The setup for a local rustc works as follows:
 ```sh
+# Clone the rust-lang/rust repo.
 git clone https://github.com/rust-lang/rust/ rustc
 cd rustc
 cp config.toml.example config.toml
 # Now edit `config.toml` and set `debug-assertions = true`.
-# This step can take 30 minutes and more.
-./x.py build src/rustc
+
+# Build a stage 2 rustc.
+# This step can take 30 minutes or more.
+./x.py build --stage 2 library/std
```

Comment:
I've just noticed that this doesn't work locally; I get an error like:

```
error[E0463]: can't find crate for `rustc_attr`
  --> src/lib.rs:14:1
   |
14 | extern crate rustc_attr;
   | ^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error

For more information about this error, try `rustc --explain E0463`.
error: could not compile `miri`
```

But `./x.py build --stage 2 compiler/rustc` fixes it even though it takes three seconds to execute. Is this an `x.py` bug?

Reply:
Wait, only three seconds to execute? Something really weird is going on here ... Maybe this is something like https://github.com/rust-lang/rust/issues/73519 ?

Reply:
Yeah `./x.py build --stage 2 compiler/rustc` is right.

Reply:
> Thus it needs a librustc and a libstd that were both built by the same rustc as the one building Miri itself. Stage 1 does not satisfy this condition, only stage 2 does.

I don't understand why rustdoc is different from miri here. Rustdoc works like this:
- beta rustc builds stage0 libstd
- beta rustc builds stage1 rustc, linking it to stage0 libstd. This works because both are built by beta.
- beta rustc builds stage1 rustdoc, linking it to stage1 rustc and stage1 libstd. This works because all three are built by beta.

Why does miri need stage2?

Reply:
Sorry, I wrote all that I know on this subject, you might have to find someone that can explain this better.^^ Maybe rustdoc does not need proc-macros to work for the build to succeed? I seem to recall those are particularly tricky.

Reply:
@Mark-Simulacrum do you have ideas?

Reply:
rustdoc and miri need not be different, they just are today for $reasons. It is not 100% clear which behavior is better; IIRC we have discussed this a number of times over the past few months and not made too much progress.

Reply:
AFAIK only the _naming_ of the stages should be different - currently `build --stage 2 src/tools/rustdoc` does the same thing as `build --stage 2 src/tools/miri` IIRC. But that's not my main question - this is building stage2 rustc itself. Why can't it use stage1 instead?

Reply:
Oh hold on, it's because the docs hard-code stage2 in the `rustup toolchain link` section :facepalm: 

@RalfJung @camelid if you use `x.py build --stage 1 library/std &&  rustup toolchain link custom build/x86_64-unknown-linux-gnu/stage1` (notice stage1, not stage2), does it work correctly?

Reply:
And if not, does `build --stage 1 compiler/rustc` work?

Reply:
Trying those now...

Reply:
`./x.py build --stage 1 library/std` didn't work:

```
error[E0463]: can't find crate for `rustc_attr`
  --> src/lib.rs:14:1
   |
14 | extern crate rustc_attr;
   | ^^^^^^^^^^^^^^^^^^^^^^^^ can't find crate

error: aborting due to previous error
```

Reply:
@jyn514 `./x.py build --stage 1 compiler/rustc` **worked**!

It took 23 minutes, so it must be doing more than `build --stage 1 library/std`.

Reply:
I just looked at the `x.py` logs and it seems like `build --stage 1 library/std` builds:

- stage0 std artifacts
- stage0 compiler artifacts
- stage1 std artifacts

But the output is kind of confusing so I'm not 100% sure.

Reply:
> ./x.py build --stage 1 compiler/rustc worked!

What exactly did you test to conclude that Miri works? Did you do `./miri test`? I definitely tried this in the past and it did not work, and IIRC proc-macros were the culprit.

Reply:
Ah no I misread, `--stage 1 compiler/rustc` builds rustc twice, so it is basically the same as `--stage 2 compiler/rustc` I think? Just some uplifting is happening.

> it seems like build --stage 1 library/std builds:

Yes. And `--stage 1 compiler/rustc` additionally builds "stage1 compiler artifacts", also known as "the stage 2 compiler".

Which switch did you use after a `./x.py build --stage 1 compiler/rustc`, to then build Miri?